File: app/src/main/java/com/elementary/tasks/core/app_widgets/events/EventsFactory.java
package com.elementary.tasks.core.app_widgets.events;

import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.support.annotation.NonNull;
import android.util.TypedValue;
import android.view.View;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.elementary.tasks.R;
import com.elementary.tasks.birthdays.BirthdayItem;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.Contacts;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.ReminderUtils;
import com.elementary.tasks.core.utils.TimeCount;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.reminder.models.Place;
import com.elementary.tasks.reminder.models.Reminder;
import com.elementary.tasks.reminder.models.ShopItem;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Copyright 2015 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class EventsFactory implements RemoteViewsService.RemoteViewsFactory {

    @NonNull
    private List<CalendarItem> data = new ArrayList<>();
    @NonNull
    private Map<String, Reminder> map = new HashMap<>();
    private Context mContext;
    private TimeCount mCount;
    private int widgetID;
    private SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());

    EventsFactory(Context ctx, Intent intent) {
        mContext = ctx;
        mCount = TimeCount.getInstance(ctx);
        widgetID = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
    }

    @Override
    public void onCreate() {
        data.clear();
        map.clear();
    }

    @Override
    public void onDataSetChanged() {
        data.clear();
        map.clear();
        boolean is24 = Prefs.getInstance(mContext).is24HourFormatEnabled();
        List<Reminder> reminderItems = RealmDb.getInstance().getEnabledReminders();
        for (Reminder item : reminderItems) {
            if (item.getViewType() == Reminder.SHOPPING) {
                continue;
            }
            int type = item.getType();
            String summary = item.getSummary();
            long eventTime = item.getDateTime();
            String id = item.getUuId();

            String time = "";
            String date = "";
            int viewType = 1;
            if (Reminder.isGpsType(type)) {
                Place place = item.getPlaces().get(0);
                date = String.format(Locale.getDefault(), "%.5f", place.getLatitude());
                time = String.format(Locale.getDefault(), "%.5f", place.getLongitude());
            } else if (Reminder.isBase(type, Reminder.BY_WEEK)) {
                Calendar calendar = Calendar.getInstance();
                calendar.setTimeInMillis(eventTime);
                date = ReminderUtils.getRepeatString(mContext, item.getWeekdays());
                time = TimeUtil.getTime(calendar.getTime(), is24);
            } else if (Reminder.isBase(type, Reminder.BY_MONTH)) {
                Calendar calendar1 = Calendar.getInstance();
                calendar1.setTimeInMillis(eventTime);
                date = TimeUtil.DATE_FORMAT.format(calendar1.getTime());
                time = TimeUtil.getTime(calendar1.getTime(), is24);
            } else if (Reminder.isSame(type, Reminder.BY_DATE_SHOP)) {
                viewType = 2;
                map.put(id, item);
            } else {
                String[] dT = mCount.getNextDateTime(eventTime);
                date = dT[0];
                time = dT[1];
            }
            data.add(new CalendarItem(CalendarItem.Type.REMINDER, summary, item.getTarget(), id, time, date, eventTime, viewType, item));
        }

        Prefs prefs = Prefs.getInstance(mContext);
        if (prefs.isBirthdayInWidgetEnabled()) {
            int mDay;
            int mMonth;
            int n = 0;
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(TimeUtil.getBirthdayTime(prefs.getBirthdayTime()));
            int hour = calendar.get(Calendar.HOUR_OF_DAY);
            int minute = calendar.get(Calendar.MINUTE);
            calendar.setTimeInMillis(System.currentTimeMillis());
            do {
                mDay = calendar.get(Calendar.DAY_OF_MONTH);
                mMonth = calendar.get(Calendar.MONTH);
                List<BirthdayItem> list = RealmDb.getInstance().getBirthdays(mDay, mMonth);
                for (BirthdayItem item : list) {
                    String birthday = item.getDate();
                    String name = item.getName();
                    long eventTime = 0;
                    try {
                        Date date = format.parse(birthday);
                        Calendar calendar1 = Calendar.getInstance();
                        calendar1.setTimeInMillis(System.currentTimeMillis());
                        int year = calendar1.get(Calendar.YEAR);
                        if (date != null) {
                            calendar1.setTime(date);
                            calendar1.set(Calendar.YEAR, year);
                            calendar1.set(Calendar.HOUR_OF_DAY, hour);
                            calendar1.set(Calendar.MINUTE, minute);
                            eventTime = calendar1.getTimeInMillis();
                        }
                    } catch (ParseException e) {
                        e.printStackTrace();
                    }
                    data.add(new CalendarItem(CalendarItem.Type.BIRTHDAY, mContext.getString(R.string.birthday), name, item.getUuId(), birthday, "", eventTime, 1, item));
                }
                calendar.setTimeInMillis(calendar.getTimeInMillis() + (1000 * 60 * 60 * 24));
                n++;
            } while (n <= 7);
        }
        Collections.sort(data, (eventsItem, o2) -> {
            long time1 = 0, time2 = 0;
            if (eventsItem.getItem() instanceof BirthdayItem) {
                BirthdayItem item = (BirthdayItem) eventsItem.getItem();
                TimeUtil.DateItem dateItem = TimeUtil.getFutureBirthdayDate(mContext, item.getDate());
                if (dateItem != null) {
                    Calendar calendar = dateItem.getCalendar();
                    time1 = calendar.getTimeInMillis();
                }
            } else if (eventsItem.getItem() instanceof Reminder) {
                Reminder reminder = (Reminder) eventsItem.getItem();
                time1 = TimeUtil.getDateTimeFromGmt(reminder.getEventTime());
            }
            if (o2.getItem() instanceof BirthdayItem) {
                BirthdayItem item = (BirthdayItem) o2.getItem();
                TimeUtil.DateItem dateItem = TimeUtil.getFutureBirthdayDate(mContext, item.getDate());
                if (dateItem != null) {
                    Calendar calendar = dateItem.getCalendar();
                    time2 = calendar.getTimeInMillis();
                }
            } else if (o2.getItem() instanceof Reminder) {
                Reminder reminder = (Reminder) o2.getItem();
                time2 = TimeUtil.getDateTimeFromGmt(reminder.getEventTime());
            }
            return (int) (time1 - time2);
        });
    }

    @Override
    public void onDestroy() {
        map.clear();
        data.clear();
    }

    @Override
    public int getCount() {
        return data.size();
    }

    @Override
    public RemoteViews getViewAt(int i) {
        SharedPreferences sp = mContext.getSharedPreferences(
                EventsWidgetConfig.EVENTS_WIDGET_PREF, Context.MODE_PRIVATE);
        int theme = sp.getInt(EventsWidgetConfig.EVENTS_WIDGET_THEME + widgetID, 0);
        EventsTheme eventsTheme = EventsTheme.getThemes(mContext).get(theme);
        int itemBackground = eventsTheme.getItemBackground();
        int itemTextColor = eventsTheme.getItemTextColor();
        float itemTextSize = sp.getFloat(EventsWidgetConfig.EVENTS_WIDGET_TEXT_SIZE + widgetID, 0);
        int checkboxColor = eventsTheme.getCheckboxColor();

        RemoteViews rView = null;
        if (i >= getCount()) {
            return null;
        }
        CalendarItem item = data.get(i);
        if (item.getViewType() == 1) {
            rView = new RemoteViews(mContext.getPackageName(), R.layout.list_item_current_widget);
            rView.setInt(R.id.itemBg, "setBackgroundResource", itemBackground);

            String task = item.getName();
            if (task == null || task.matches("")) {
                task = Contacts.getNameFromNumber(item.getNumber(), mContext);
            }
            rView.setTextViewText(R.id.taskText, task);
            rView.setTextColor(R.id.taskText, itemTextColor);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                rView.setTextViewTextSize(R.id.taskText, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
                rView.setTextViewTextSize(R.id.taskNumber, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
                rView.setTextViewTextSize(R.id.taskDate, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
                rView.setTextViewTextSize(R.id.taskTime, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
                rView.setTextViewTextSize(R.id.leftTime, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
            } else {
                rView.setFloat(R.id.taskTime, "setTextSize", itemTextSize);
                rView.setFloat(R.id.taskDate, "setTextSize", itemTextSize);
                rView.setFloat(R.id.taskNumber, "setTextSize", itemTextSize);
                rView.setFloat(R.id.taskText, "setTextSize", itemTextSize);
                rView.setFloat(R.id.leftTime, "setTextSize", itemTextSize);
            }

            String number = item.getNumber();
            if (number != null && !number.matches("")) {
                rView.setTextViewText(R.id.taskNumber, number);
                rView.setTextColor(R.id.taskNumber, itemTextColor);
            } else {
                rView.setViewVisibility(R.id.taskNumber, View.GONE);
            }
            rView.setTextViewText(R.id.taskDate, item.getDayDate());
            rView.setTextColor(R.id.taskDate, itemTextColor);

            rView.setTextViewText(R.id.taskTime, item.getTime());
            rView.setTextColor(R.id.taskTime, itemTextColor);

            rView.setTextViewText(R.id.leftTime, mCount.getRemaining(item.getDate()));
            rView.setTextColor(R.id.leftTime, itemTextColor);

            if (item.getId() != null) {
                Intent fillInIntent = new Intent();
                fillInIntent.putExtra(Constants.INTENT_ID, item.getId());
                if (item.getType() == CalendarItem.Type.REMINDER) {
                    fillInIntent.putExtra(EventEditService.TYPE, true);
                } else {
                    fillInIntent.putExtra(EventEditService.TYPE, false);
                }
                rView.setOnClickFillInIntent(R.id.taskDate, fillInIntent);
                rView.setOnClickFillInIntent(R.id.taskTime, fillInIntent);
                rView.setOnClickFillInIntent(R.id.taskNumber, fillInIntent);
                rView.setOnClickFillInIntent(R.id.taskText, fillInIntent);
                rView.setOnClickFillInIntent(R.id.itemBg, fillInIntent);
            }
        }
        if (item.getViewType() == 2) {
            rView = new RemoteViews(mContext.getPackageName(),
                    R.layout.list_item_current_widget_with_list);
            rView.setInt(R.id.itemBg, "setBackgroundResource", itemBackground);
            String task = item.getName();
            rView.setTextViewText(R.id.taskText, task);
            rView.setTextColor(R.id.taskText, itemTextColor);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                rView.setTextViewTextSize(R.id.taskText, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
            } else {
                rView.setFloat(R.id.taskText, "setTextSize", itemTextSize);
            }

            int count = 0;
            List<ShopItem> lists = map.get(item.getId()).getShoppings();
            rView.removeAllViews(R.id.todoList);
            for (ShopItem list : lists) {
                RemoteViews view = new RemoteViews(mContext.getPackageName(), R.layout.list_item_task_item_widget);
                boolean isBlack = checkboxColor == 0;
                if (list.isChecked()) {
                    if (isBlack) {
                        view.setInt(R.id.checkView, "setBackgroundResource", R.drawable.ic_check_box_black_24dp);
                    } else {
                        view.setInt(R.id.checkView, "setBackgroundResource", R.drawable.ic_check_box_white_24dp);
                    }
                } else {
                    if (isBlack) {
                        view.setInt(R.id.checkView, "setBackgroundResource", R.drawable.ic_check_box_outline_blank_black_24dp);
                    } else {
                        view.setInt(R.id.checkView, "setBackgroundResource", R.drawable.ic_check_box_outline_blank_white_24dp);
                    }
                }

                view.setTextColor(R.id.shopText, itemTextColor);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    view.setTextViewTextSize(R.id.shopText, TypedValue.COMPLEX_UNIT_SP, itemTextSize);
                } else {
                    view.setFloat(R.id.shopText, "setTextSize", itemTextSize);
                }

                count++;
                if (count == 9) {
                    view.setViewVisibility(R.id.checkView, View.INVISIBLE);
                    view.setTextViewText(R.id.shopText, "...");
                    rView.addView(R.id.todoList, view);
                    break;
                } else {
                    view.setViewVisibility(R.id.checkView, View.VISIBLE);
                    view.setTextViewText(R.id.shopText, list.getSummary());
                    rView.addView(R.id.todoList, view);
                }
            }

            Intent fillInIntent = new Intent();
            fillInIntent.putExtra(Constants.INTENT_ID, item.getId());
            fillInIntent.putExtra(EventEditService.TYPE, true);
            rView.setOnClickFillInIntent(R.id.taskText, fillInIntent);
            rView.setOnClickFillInIntent(R.id.itemBg, fillInIntent);
            rView.setOnClickFillInIntent(R.id.todoList, fillInIntent);
        }
        return rView;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 2;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }
}

File: app/src/main/java/com/elementary/tasks/core/app_widgets/events/EventsFactory.kt
package com.elementary.tasks.core.app_widgets.events

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.util.TypedValue
import android.view.View
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import androidx.annotation.ColorInt
import androidx.core.content.ContextCompat
import com.elementary.tasks.R
import com.elementary.tasks.core.app_widgets.AppWidgetActionActivity
import com.elementary.tasks.core.app_widgets.Direction
import com.elementary.tasks.core.app_widgets.WidgetIntentProtocol
import com.elementary.tasks.core.app_widgets.WidgetUtils
import com.elementary.tasks.core.data.adapter.birthday.UiBirthdayWidgetListAdapter
import com.elementary.tasks.core.data.adapter.reminder.UiReminderWidgetListAdapter
import com.elementary.tasks.core.data.repository.BirthdayRepository
import com.elementary.tasks.core.data.repository.ReminderRepository
import com.elementary.tasks.core.data.ui.birthday.UiBirthdayWidgetList
import com.elementary.tasks.core.data.ui.reminder.widget.UiReminderWidgetList
import com.elementary.tasks.core.data.ui.reminder.widget.UiReminderWidgetShopList
import com.elementary.tasks.core.data.ui.widget.DateSorted
import com.elementary.tasks.core.utils.Constants
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.elementary.tasks.core.utils.ui.ViewUtils
import com.elementary.tasks.core.utils.ui.font.FontParams

class EventsFactory(
  private val context: Context,
  intent: Intent,
  private val prefs: Prefs,
  private val dateTimeManager: DateTimeManager,
  private val reminderRepository: ReminderRepository,
  private val birthdayRepository: BirthdayRepository,
  private val uiReminderWidgetListAdapter: UiReminderWidgetListAdapter,
  private val uiBirthdayWidgetListAdapter: UiBirthdayWidgetListAdapter
) : RemoteViewsService.RemoteViewsFactory {

  private var data = mutableListOf<DateSorted>()
  private var itemTextSize: Float = FontParams.DEFAULT_FONT_SIZE.toFloat()
  private var itemBg: Int = 0
  @ColorInt
  private var textColor: Int = Color.BLACK

  private val widgetID: Int = intent.getIntExtra(
    AppWidgetManager.EXTRA_APPWIDGET_ID,
    AppWidgetManager.INVALID_APPWIDGET_ID
  )
  private val prefsProvider = EventsWidgetPrefsProvider(context, widgetID)

  override fun onCreate() {
    data.clear()
  }

  override fun onDataSetChanged() {
    data.clear()

    itemTextSize = prefsProvider.getTextSize()
    val itemBgColor = prefsProvider.getItemBackground()

    itemBg = WidgetUtils.newWidgetBg(itemBgColor)

    textColor = if (WidgetUtils.isDarkBg(itemBgColor)) {
      ContextCompat.getColor(context, R.color.pureWhite)
    } else {
      ContextCompat.getColor(context, R.color.pureBlack)
    }

    reminderRepository.getActive()
      .map { uiReminderWidgetListAdapter.create(it, textColor) }
      .also { data.addAll(it) }

    if (prefs.isBirthdayInWidgetEnabled) {
      val dateTime = dateTimeManager.getCurrentDateTime()

      birthdayRepository.getByDayMonth(dateTime.dayOfMonth, dateTime.monthValue - 1)
        .map { uiBirthdayWidgetListAdapter.convert(it) }
        .also { data.addAll(it) }
    }

    data = data.sortedWith(compareBy { it.millis }).toMutableList()
  }

  override fun onDestroy() {
    data.clear()
  }

  override fun getCount(): Int {
    return data.size
  }

  override fun getViewAt(i: Int): RemoteViews? {
    var rv: RemoteViews? = null
    if (i >= count) {
      return null
    }

    when (val item = data[i]) {
      is UiBirthdayWidgetList -> {
        val icon = ViewUtils.createIcon(context, R.drawable.ic_twotone_cake_24px, textColor)

        rv = RemoteViews(context.packageName, R.layout.list_item_widget_birthday)

        rv.setInt(R.id.itemBackgroundView, "setBackgroundResource", itemBg)

        rv.setTextColor(R.id.nameView, textColor)
        rv.setTextColor(R.id.ageBirthDateView, textColor)
        rv.setTextColor(R.id.leftTimeView, textColor)

        rv.setImageViewBitmap(R.id.statusIconView, icon)

        rv.setTextViewText(R.id.nameView, item.name)
        rv.setTextViewText(R.id.ageBirthDateView, item.ageFormattedAndBirthdayDate)
        rv.setTextViewText(R.id.leftTimeView, item.remainingTimeFormatted)

        val data = WidgetIntentProtocol(
          mapOf<String, Any?>(
            Pair(Constants.INTENT_ID, item.uuId)
          )
        )

        val fillInIntent = Intent()
        fillInIntent.putExtra(AppWidgetActionActivity.DATA, data)
        fillInIntent.putExtra(AppWidgetActionActivity.DIRECTION, Direction.BIRTHDAY)
        rv.setOnClickFillInIntent(R.id.itemBackgroundView, fillInIntent)
      }
      is UiReminderWidgetShopList -> {
        val icon = ViewUtils.createIcon(context, R.drawable.ic_twotone_shopping_cart_24px, textColor)

        rv = RemoteViews(context.packageName, R.layout.list_item_widget_shop_list)
        rv.setInt(R.id.itemBackgroundView, "setBackgroundResource", itemBg)

        rv.setImageViewBitmap(R.id.statusIconView, icon)

        rv.setTextViewText(R.id.taskTextView, item.text)
        rv.setTextColor(R.id.taskTextView, textColor)
        rv.setTextViewTextSize(R.id.taskTextView, TypedValue.COMPLEX_UNIT_SP, itemTextSize)

        if (!item.dateTime.isNullOrEmpty()) {
          rv.setTextViewText(R.id.dateTimeView, item.dateTime)
          rv.setViewVisibility(R.id.dateTimeView, View.VISIBLE)
        } else {
          rv.setViewVisibility(R.id.dateTimeView, View.GONE)
        }

        var count = 0

        rv.removeAllViews(R.id.todoList)

        for (task in item.items) {
          val view = RemoteViews(context.packageName, R.layout.list_item_widget_shop_item)

          view.setImageViewBitmap(R.id.checkView, task.icon)
          view.setTextColor(R.id.shopText, textColor)
          view.setTextViewTextSize(R.id.shopText, TypedValue.COMPLEX_UNIT_SP, itemTextSize)

          count++

          if (count == 9) {
            view.setViewVisibility(R.id.checkView, View.INVISIBLE)
            view.setTextViewText(R.id.shopText, "...")
            rv.addView(R.id.todoList, view)
            break
          } else {
            view.setViewVisibility(R.id.checkView, View.VISIBLE)
            view.setTextViewText(R.id.shopText, task.text)
            rv.addView(R.id.todoList, view)
          }
        }

        val data = WidgetIntentProtocol(
          mapOf<String, Any?>(
            Pair(Constants.INTENT_ID, item.uuId)
          )
        )

        val fillInIntent = Intent()
        fillInIntent.putExtra(AppWidgetActionActivity.DATA, data)
        fillInIntent.putExtra(AppWidgetActionActivity.DIRECTION, Direction.REMINDER)
        rv.setOnClickFillInIntent(R.id.taskTextView, fillInIntent)
        rv.setOnClickFillInIntent(R.id.itemBackgroundView, fillInIntent)
        rv.setOnClickFillInIntent(R.id.todoList, fillInIntent)
      }
      is UiReminderWidgetList -> {
        val icon = ViewUtils.createIcon(context, R.drawable.ic_twotone_alarm_24px, textColor)

        rv = RemoteViews(context.packageName, R.layout.list_item_widget_events)

        rv.setInt(R.id.itemBackgroundView, "setBackgroundResource", itemBg)

        rv.setTextColor(R.id.taskTextView, textColor)
        rv.setTextColor(R.id.dateTimeView, textColor)
        rv.setTextColor(R.id.leftTimeView, textColor)

        rv.setTextViewTextSize(R.id.taskTextView, TypedValue.COMPLEX_UNIT_SP, itemTextSize)
        rv.setTextViewTextSize(R.id.dateTimeView, TypedValue.COMPLEX_UNIT_SP, itemTextSize)
        rv.setTextViewTextSize(R.id.leftTimeView, TypedValue.COMPLEX_UNIT_SP, itemTextSize)

        rv.setImageViewBitmap(R.id.statusIconView, icon)
        rv.setTextViewText(R.id.taskTextView, item.text)
        rv.setTextViewText(R.id.dateTimeView, item.dateTime)
        rv.setTextViewText(R.id.leftTimeView, item.remainingTimeFormatted)

        val data = WidgetIntentProtocol(
          mapOf<String, Any?>(
            Pair(Constants.INTENT_ID, item.uuId)
          )
        )

        val fillInIntent = Intent()
        fillInIntent.putExtra(AppWidgetActionActivity.DATA, data)
        fillInIntent.putExtra(AppWidgetActionActivity.DIRECTION, Direction.REMINDER)
        rv.setOnClickFillInIntent(R.id.taskTextView, fillInIntent)
        rv.setOnClickFillInIntent(R.id.itemBackgroundView, fillInIntent)
      }
    }
    return rv
  }

  override fun getLoadingView(): RemoteViews? {
    return null
  }

  override fun getViewTypeCount(): Int {
    return 3
  }

  override fun getItemId(position: Int): Long {
    return position.toLong()
  }

  override fun hasStableIds(): Boolean {
    return true
  }
}
